Cause collect_output_until_deadline to return when session exits.

CarloWood opened the pull request:
The `yield_time_ms` parameter of the model issued exec_command
and `write_stdin ToolCalls` is meant as a safe-guard: a deadline
for the maximum time to wait for shell output.

> Note: in the case of `write_stdin` that only really makes sense if
> the `write_stdin` issued a command itself and isn't a poll-- but the
> test `suite::unified_exec::unified_exec_timeout_and_followup_poll`,
> that was recently added, tests that even in the case of a poll it
> waits (for the output of a previous exec_command), so I didn't
> change this. The model typically sends a yield_time_ms of 1000
> with every poll, which is OK.

If there is no output expected anymore at all, because the process
exited, then that is "output" too: namely, the process telling the
CLI that no more output will follow (`completed`).

`collect_output_until_deadline` should not continue waiting for a
notification that there is more output to be read if the PTY closed
that connection: it is done collecting output and should return.

This patch fixes the fact that `collect_output_until_deadline` would
wait till the end of `deadline`, even after the reader task saw EOF.

# Description of changes

When the reader task (`reader_handle`) exits because it reached EOF or
because of a fatal read error, it sends an empty message to `output_tx`.
This guarantees that this is synchronized with (happens after) any
output written by the PTY.

When `output_task` receives the empty chunk, it marks the output (buffer) as
`completed`-- notifies any waiters, waking up `collect_output_until_deadline`
from a possible `tokio::select!`-- and exits.

If `collect_output_until_deadline` sees that the output buffer has the `completed`
flag set, then it exits immediately (as opposed to waiting till the deadline
is reached, as was the case before this patch).

By using `completed` instead of `has_exited`, there is no race and it is
guaranteed that `collect_output_until_deadline` does not return until *all*
output has been collected (unless canceled).

Both `exec_command` and `write_stdin` - that call `collect_output_until_deadline`,
if they see that `output_completed` is set, follow up with a call to
`wait_for_exited_until_deadline` (where the `deadline` is possibly extended to
be at least 1 second into the future at this point), to wait for the process
to actually exit. If the process exits they continue immediately, otherwise,
if the process does not exit within the set deadline despite that the PTY closed
its output fd(s) they assume the process is still "Alive".

Related issue link: [6715](https://github.com/openai/codex/issues/6715)

CarloWood commented:
I have read the CLA Document and I hereby sign the CLA

etraut-openai commented:
Thanks for the contribution. Is there a bug report associated with this PR? I don't see a link above. We prefer to have bug reports to each PR. It helps us track, prioritize, and assess risks.

CarloWood committed:
Cause collect_output_until_deadline to return when session exits.

The `yield_time_ms` parameter of the model issued exec_command
and write_stdin ToolCalls is meant as a safe-guard: a deadline
for the maximum time to wait for shell output.

[Note: in the case of write_stdin that only really makes sense if
the write_stdin issued a command itself and isn't a poll-- but the
test `suite::unified_exec::unified_exec_timeout_and_followup_poll`,
that was recently added, tests that even in the case of a poll it
waits (for the output of a previous exec_command), so I didn't
change this. The model typically sends a yield_time_ms of 1000
with every poll, which is OK.]

If there is no output expected anymore at all, because the process
exited, then that is "output" too: namely, the process telling the
CLI that no more output will follow (`completed`).

collect_output_until_deadline should not continue waiting for a
notification that there is more output to be read if the PTY closed
that connection: it is done collecting output and should return.

This patch fixes the fact that collect_output_until_deadline would
wait till the end of `deadline`, even after the reader task saw EOF.

Implementation:

When the reader task (reader_handle) exits because it reached EOF or
because of a fatal read error, it sends an empty message to output_tx.
This guarantees that this is synchronized with (happens after) any
output written by the PTY.

When output_task receives the empty chunk, it marks the output (buffer) as
`completed`-- notifies any waiters, waking up collect_output_until_deadline
from a possible tokio::select!-- and exits.

If collect_output_until_deadline sees that the output buffer has the `completed`
flag set, then it exits immediately (as opposed to waiting till the deadline
is reached, as was the case before this patch).

By using `completed` instead of `has_exited`, there is no race and it is
guaranteed that collect_output_until_deadline does not return until *all*
output has been collected.

Both, exec_command and write_stdin, that call collect_output_until_deadline,
if they see that `output_completed` is set, follow up with a call to
wait_for_exited_until_deadline (where the deadline is possibly extended to
be at least 1 second into the future at this point), to wait for the process
to actually exit. If the process exits they continue immediately, otherwise,
if the process does not exit the set deadline, despite that the PTY closed
its output fd(s), they assume the process is still "Alive".

CarloWood commented on codex-rs/core/src/unified_exec/session_manager.rs:
```diff
@@ -481,6 +497,9 @@ impl UnifiedExecSessionManager {
+                if session.is_some_and(UnifiedExecSession::has_exited) {
+                    break;
+                }
                 let remaining = deadline.saturating_duration_since(Instant::now());
```
👍
I guess that model is a lot better than gpt-5.1-codex under my plan.
Anyway, I basically rewrote this - so I did a forced push.

CarloWood commented:
> Thanks for the contribution. Is there a bug report associated with this PR? I don't see a link above. We prefer to have bug reports to each PR. It helps us track, prioritize, and assess risks.

Yes, this one: https://github.com/openai/codex/issues/6715

etraut-openai commented:
Thanks for the contribution. Before I assign this to someone on the codex team to review, please fix the lint (cargo clippy) errors.

CarloWood commented:
I added a new test: unified_exec_stops_waiting_after_process_exit
To run it:
```
cd codex-rs
cargo test -p codex-core unified_exec_stops_waiting_after_process_exit -- --nocapture
```

CarloWood committed:
Add test suite::unified_exec::unified_exec_stops_waiting_after_process_exit

This test spawns an exec_command with yield_time_ms = 5000:

    perl -e 'print "start\n"; select undef, undef, undef, 0.25; print "end\n";'

Here is what happens (while I still had debug lines in the code):

10.332100Z  INFO UnifiedExecSession: got new chunk, pushing it to output buffer
10.332171Z  INFO UnifiedExecSession: notifying waiters
10.336221Z  INFO collect_output_until_deadline: top of loop
10.336265Z  INFO collect_output_until_deadline: chunk="start\r\n"
10.336303Z  INFO collect_output_until_deadline: top of loop
10.336324Z  INFO collect_output_until_deadline: drained_chunks.is_empty
10.336345Z  INFO collect_output_until_deadline: remaining: 4.999857405s (entering sleep!)
<--250 ms delay-->
10.582412Z  INFO UnifiedExecSession: got new chunk, pushing it to output buffer
10.582493Z  INFO UnifiedExecSession: notifying waiters
sleep! wakes up
10.582570Z  INFO collect_output_until_deadline: top of loop
10.582616Z  INFO collect_output_until_deadline: chunk="end\r\n"
10.582655Z  INFO collect_output_until_deadline: top of loop
10.582693Z  INFO collect_output_until_deadline: drained_chunks.is_empty
10.582721Z  INFO collect_output_until_deadline: remaining: 4.753480629s (entering sleep!)
The fix (see next commit):
10.582715Z  INFO UnifiedExecSession: Ok: chunk.is_empty(), calling mark_completed()
10.582763Z  INFO UnifiedExecSession: notifying waiters
10.582793Z  INFO UnifiedExecSession: reader task exited
sleep! wakes up
10.582893Z  INFO collect_output_until_deadline: top of loop
10.582926Z  INFO collect_output_until_deadline: drained_chunks.is_empty
10.582952Z  INFO collect_output_until_deadline: completed!

In English: the script write "start" to the stdout which is read
by the reader task and pushed into the output buffer.

collect_output_until_deadline reads it; attempts to read more
but there is no more (drained_chunks.is_empty) and enters sleep! with
4.999857405s remaining.

The reader task reads "end" 250 ms later and pushes that to the
output buffer.

collect_output_until_deadline reads it; attempts to read more
but there is no more (drained_chunks.is_empty) and enters sleep! with
4.753480629s remaining.

WITHOUT THE COMMIT THAT FOLLOWS THIS COMMIT: collect_output_until_deadline
hangs for the full 4.75s, until the deadline, and only then releases
the read output.

WITH THE FIX:

The fact that the PTYs output is closed is detected and an empty
chunk is written to the tx_output channel. This causes the 'completed'
flag to be set on the output buffer.

collect_output_until_deadline wakes up because of the notify,
tries to read from the buffer and is told that the state of the buffer
is 'completed' (no more output follows), so it exits immediately.

CarloWood commented:
Did a new rebase (and git push --force) to satisfy the demand from `contributing.md`:
Keep commits atomic. Each commit should compile and the tests should pass. This makes reviews and potential rollbacks easier.

The PR now exists of two commits: first a new test is added (that fails!), the second commit then fixes the bug so that the test succeeds.

CarloWood commented:
The Lint problems seems to be on your end:

error: could not execute process `sccache C:\Users\runneradmin\.rustup\toolchains\1.90.0-x86_64-pc-windows-msvc\bin\clippy-driver.exe C:\Users\runneradmin\.rustup\toolchains\1.90.0-x86_64-pc-windows-msvc\bin\rustc.exe -vV` (never executed)

CarloWood commented:
Note: the test that fail on `Tests — ubuntu-24.04 - x86_64-unknown-linux-gnu` here, succeed on my box. I don't know why they fail in rust-ci :/.

etraut-openai commented:
OK, sounds good. I will move it out of draft status once we fix our failing CI nodes and confirm that the status is green.

etraut-openai commented:
> Do you want me to try to fix this (while I can't reproduce the problem)?

Thanks for the analysis. I think we should fix this separately.

CarloWood commented:
I fixed two race conditions that surfaced (see last two commits).
I also changed the main commit because of the need for better exposure of the `completed` boolean. I made it an atomic and added accessors that can be used to get rid of the race conditions: call session.finished() instead of session.has_exited().

Five more tests now fail due to adding a sleep of 200ms in front of setting exit_status
to true:

failures:
    suite::approvals::approval_matrix_covers_all_modes
    suite::unified_exec::exec_command_reports_chunk_and_exit_metadata
    suite::unified_exec::unified_exec_emits_end_event_when_session_dies_via_stdin
    suite::unified_exec::unified_exec_emits_exec_command_end_event
    suite::unified_exec::write_stdin_returns_exit_metadata_and_clears_session

I will have a look at those tomorrow.

CarloWood commented:
Ignore that last forced push please... I think I accidentally included some debug code.

CarloWood commented on codex-rs/core/src/unified_exec/session_manager.rs:
```diff
@@ -206,6 +228,22 @@ impl UnifiedExecSessionManager {
+            let deadline = now + remaining;
+            if let Some(entry) = self.sessions.lock().await.get(&session_id) {
+                Self::wait_for_exited_until_deadline(&entry.session, deadline).await;
+            }
```
👍

CarloWood committed:
Release sessions lock before waiting for exit signal.

This separate commit addresses https://github.com/openai/codex/pull/6718#discussion_r2548248412

It combines exit_notify: Arc<Notify> and exit_status: Arc<AtomicBool>
into exit_status: Arc<ExitStatus> where ExitStatus is a new object
that handles threadsafe signaling and monitoring of the exit status
flag.

If the PTY process (wait_handle) observes that the child process
did exit then it calls ExitStatus::signal().

Polling is possible using ExitStatus::signal_received(), but that
is a fuzzy boolean (True or WasFalse): if the function returns
false then that only means it was very recently false.

One can wait for the signal to be received with
`ExitStatus::wait_for_exit_until(deadline)`, which returns true
if the process did exit and false if it didn't (but the deadline
was reached).

Note: the cheap and dirty "solution" to just drop the guard before
calling `wait_for_exited_until_deadline` meant that there would be
a race between the last check on the `exited` boolean and the
creation of the notify.notified() future, during which there was
a tiny window where we could miss the exit notify and wait till the
dealine for nothing. Wrapping all of this into its own class is much
cleaner and now we don't have such race.

etraut-openai commented:
I hear you. I appreciate all the work you put in, but we're unlikely to revert the existing fix.

The codex code base evolves rapidly. We've been averaging about 40 commits a day, so keeping any PR fresh is a challenge. Keeping up with all of the merges is also difficult. I just learned that my colleague had introduced a fix for this issue a few days ago.

When we accept PRs from the community, we need to be sure that they don't cause regressions or maintainability problems, so it takes quite a bit of work on our part to review, iterate on, and accept community PRs. It's often more efficient for us to implement the fix from scratch since we have more context about the code, etc. I hope that makes sense.

I presume that you posted this bug report and invested in a fix because this was a pain point that you were personally hitting. Hopefully the most recent published CLI addresses the problem.
